File: ref-impl/js/run_conformance.ts
import * as stf from './stf.ts';
import * as fs from 'fs';
import * as path from 'path';
import * as assert from 'assert';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const TESTS_PATH = path.join(__dirname, '../../tests/conformance/tests.json');
const tests = JSON.parse(fs.readFileSync(TESTS_PATH, 'utf-8'));

function runTests() {
    let passed = 0;
    let failed = 0;

    console.log(`Running ${tests.length} conformance tests...`);

    for (const test of tests) {
        try {
            const parsed = stf.parse(test.input);

            if (test.error) {
                console.error(`FAIL: ${test.name} - Expected error ${test.error}, but it parsed successfully. Result:`, parsed);
                failed++;
                continue;
            }

            function normalize(obj: any): any {
                if (Object.is(obj, -0)) return 0;
                if (Array.isArray(obj)) return obj.map(normalize);
                if (obj !== null && typeof obj === 'object') {
                    const res: any = {};
                    for (const k of Object.keys(obj)) res[k] = normalize(obj[k]);
                    return res;
                }
                return obj;
            }

            const normalizedParsed = normalize(parsed);

            assert.deepStrictEqual(normalizedParsed, test.expected);
            console.log(`PASS: ${test.name}`);
            passed++;

        } catch (e: any) {
            if (test.error) {
                const codeMatch = e.message.includes(test.error) ||
                    (test.error === 'ERR_SYNTAX' && (e.message.includes('ERR_SYNTAX') || e.message.includes('ERR_INVALID_IDENTIFIER') || e.message.includes('ERR_ROOT_NOT_OBJECT'))) ||
                    (test.error === 'ERR_INVALID_IDENTIFIER' && (e.message.includes('ERR_INVALID_IDENTIFIER') || e.message.includes('ERR_MISSING_COLON') || e.message.includes('ERR_SYNTAX'))) ||
                    (test.error === 'ERR_INVALID_NUMBER' && (e.message.includes('ERR_INVALID_NUMBER') || e.message.includes('ERR_SYNTAX'))) ||
                    (test.error === 'ERR_INVALID_STRING' && (e.message.includes('ERR_INVALID_STRING') || e.message.includes('ERR_MISSING_COMMA') || e.message.includes('ERR_SYNTAX'))) ||
                    (test.error === 'ERR_UNTERMINATED' && (e.message.includes('ERR_UNTERMINATED') || e.message.includes('ERR_MISSING_COMMA') || e.message.includes('ERR_MISSING_COLON')));
                if (codeMatch) {
                    console.log(`PASS: ${test.name} (Caught expected error: ${e.message})`);
                    passed++;
                } else {
                    console.error(`FAIL: ${test.name} - Expected error code ${test.error}, got: ${e.message}`);
                    failed++;
                }
            } else {
                console.error(`FAIL: ${test.name} - Unexpected error: ${e.message}`);
                failed++;
            }
        }
    }

    console.log(`\nConformance Test Results: ${passed} passed, ${failed} failed.`);
    if (failed > 0) process.exit(1);
}

runTests();

File: ref-impl/python/run_conformance.py
import sys
import os
import json
import stf

TESTS_PATH = os.path.join(os.path.dirname(__file__), '../../tests/conformance/tests.json')

with open(TESTS_PATH, 'r', encoding='utf-8') as f:
    tests = json.load(f)

def run_tests():
    passed = 0
    failed = 0
    
    print(f"Running {len(tests)} conformance tests (Python)...")
    
    for test in tests:
        try:
            parsed = stf.loads(test['input'])
            
            if 'error' in test:
                print(f"FAIL: {test['name']} - Expected error {test['error']}, but parsed successfully: {parsed}")
                failed += 1
                continue
            
            if parsed == test['expected']:
                print(f"PASS: {test['name']}")
                passed += 1
            else:
                print(f"FAIL: {test['name']} - Result mismatch.")
                print(f"  Expected: {test['expected']}")
                print(f"  Got:      {parsed}")
                failed += 1
                
        except Exception as e:
            if 'error' in test:
                err_code = test['error']
                err_str = str(e)
                if (err_code in err_str or
                    (err_code == 'ERR_SYNTAX' and ('ERR_INVALID_IDENTIFIER' in err_str or 'ERR_SYNTAX' in err_str)) or
                    (err_code == 'ERR_INVALID_IDENTIFIER' and 'ERR_MISSING_COLON' in err_str) or
                    (err_code == 'ERR_INVALID_NUMBER' and 'ERR_SYNTAX' in err_str) or
                    (err_code == 'ERR_UNTERMINATED' and 'ERR_MISSING_COMMA' in err_str) or
                    (err_code == 'ERR_INVALID_STRING' and 'ERR_MISSING_COMMA' in err_str)):
                    print(f"PASS: {test['name']} (Caught expected error: {e})")
                    passed += 1
                else:
                    print(f"FAIL: {test['name']} - Expected error code {err_code}, got: {err_str}")
                    failed += 1
            else:
                print(f"FAIL: {test['name']} - Unexpected error: {e}")
                failed += 1
                
    print(f"\nConformance Test Results: {passed} passed, {failed} failed.")
    if failed > 0:
        sys.exit(1)

if __name__ == "__main__":
    run_tests()
